src/oid4vp/holder.rs

```diff
@@ -0,0 +1,511 @@
+use super::error::OID4VPError;
+use super::permission_request::*;
+use crate::common::*;
+use crate::credential::*;
+use crate::vdc_collection::VdcCollection;
+
+use std::sync::Arc;
+
+use futures::StreamExt;
+use oid4vp::core::authorization_request::parameters::ClientIdScheme;
+use oid4vp::core::credential_format::ClaimFormatDesignation;
+use oid4vp::core::presentation_definition::PresentationDefinition;
+use oid4vp::{
+    core::{
+        authorization_request::{
+            parameters::ResponseMode,
+            verification::{did::verify_with_resolver, RequestVerifier},
+            AuthorizationRequestObject,
+        },
+        metadata::WalletMetadata,
+        presentation_submission::{DescriptorMap, PresentationSubmission},
+        response::{parameters::VpToken, AuthorizationResponse, UnencodedAuthorizationResponse},
+    },
+    // verifier::request_signer::RequestSigner,
+    wallet::Wallet as OID4VPWallet,
+};
+// use ssi::claims::jws::JwsSigner;
+use ssi::dids::DIDWeb;
+use ssi::dids::VerificationMethodDIDResolver;
+use ssi::prelude::AnyJwkMethod;
+// use ssi::JWK;
+use tokio::sync::RwLock;
+use uniffi::deps::{anyhow, log};
+
+/// A Holder is an entity that possesses one or more Verifiable Credentials.
+/// The Holder is typically the subject of the credentials, but not always.
+/// The Holder has the ability to generate Verifiable Presentations from
+/// these credentials and share them with Verifiers.
+#[derive(Debug, uniffi::Object)]
+pub struct Holder {
+    /// An atomic reference to the VDC collection.
+    pub(crate) vdc_collection: Option<Arc<VdcCollection>>,
+
+    /// Metadata about the holder.
+    pub(crate) metadata: WalletMetadata,
+
+    /// HTTP Request Client
+    pub(crate) client: oid4vp::core::util::ReqwestClient,
+
+    /// A list of trusted DIDs.
+    pub(crate) trusted_dids: Vec<String>,
+
+    // /// A request signer callback interface to handle
+    // /// cryptographic signing with native libraries.
+    // pub(crate) request_signer: Option<Arc<dyn RequestSignerInterface>>,
+    /// Provide optional credentials to the holder instance.
+    pub(crate) provided_credentials: Option<Vec<Arc<Credential>>>,
```

Comment:
Can you please use `ParsedCredential` rather than `Credential`? The `Credential` is a generic type for storage, but we always want to operate on `ParsedCredential`s.
```suggestion
    /// Provide optional credentials to the holder instance.
    pub(crate) provided_credentials: Option<Vec<Arc<ParsedCredential>>>,
```